### PR TITLE
Native LwIP support update

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -2120,6 +2120,9 @@ int InitSSL_Ctx(WOLFSSL_CTX* ctx, WOLFSSL_METHOD* method, void* heap)
 #elif defined(WOLFSSL_APACHE_MYNEWT) && !defined(WOLFSSL_LWIP)
     ctx->CBIORecv = Mynewt_Receive;
     ctx->CBIOSend = Mynewt_Send;
+#elif defined WOLFSSL_LWIP_NATIVE
+		ctx->CBIORecv = LwIPNativeReceive;
```
Comment (dgarske) [inline src/internal.c:2124]: Check indent. space vs tabs.

File: src/wolfio.c [hunk 2]
```diff
@@ -2552,4 +2552,151 @@ int GNRC_GenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
...
+
+int LwIPNativeReceive(WOLFSSL* ssl, char* buf, int sz,
+                                     void* ctx)
+{
+    struct pbuf *current, *head;
+    WOLFSSL_LWIP_NATIVE_STATE* nlwip;
+    int ret = 0;
+
+    if (nlwip == NULL || ctx == NULL) {
+            return WOLFSSL_CBIO_ERR_GENERAL;
```
Comment (dgarske) [inline src/wolfio.c:2578]: Also indent here

File: src/wolfio.c [hunk 3]
```diff
@@ -2552,4 +2552,151 @@ int GNRC_GenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
...
+
+            if (read + len > sz) {
+                /* should never be hit but have sanity check before use */
+                return WOLFSSL_CBIO_ERR_GENERAL;
+            }
+
+            /* check if is a partial read from before */
+            XMEMCPY(&buf[read],
+                   (const char *)&(((char *)(current->payload))[nlwip->pulled]),
+
```
Comment (dgarske) [inline src/wolfio.c:2606]: Extra space or line length, can't tell

File: src/wolfio.c [hunk 4]
```diff
@@ -2552,4 +2552,151 @@ int GNRC_GenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
...
+    ssl->lwipCtx.sent_fn = sent_fn; /*  sent user callback */
+    ssl->lwipCtx.arg  = arg;
+    ssl->lwipCtx.pbuf = 0;
+    ssl->lwipCtx.pulled = 0;
+    ssl->lwipCtx.wait   = 0;
+
+    /* wolfSSL_LwIP_recv/sent_cb invokes recv/sent user callback in them. */
+    tcp_recv(pcb, LwIPNativeReceiveCB);
+    tcp_sent(pcb, LwIPNativeSentCB);
+    tcp_arg (pcb, (void *)&(ssl->lwipCtx));
```
Comment (dgarske) [inline src/wolfio.c:2694]: Prefer written as `(void*)&ssl->lwipCtx`

File: wolfcrypt/src/misc.c [hunk 5]
```diff
@@ -422,17 +422,18 @@ WC_STATIC WC_INLINE word32 btoi(byte b)
...
-{
-    if (ch >= '0' && ch <= '9')
-        ch -= '0';
-    else if (ch >= 'A' && ch <= 'F')
-        ch -= 'A' - 10;
-    else if (ch >= 'a' && ch <= 'f')
-        ch -= 'a' - 10;
+WC_STATIC WC_INLINE signed char HexCharToByte(char ch)
+{
+    signed char ret = (signed char)ch;
```
Comment (dgarske) [inline wolfcrypt/src/misc.c:427]: Cool, thanks for the explicit signed char work here.

File: wolfcrypt/src/wc_port.c [hunk 6]
```diff
@@ -2476,6 +2476,33 @@ time_t time(time_t * timer)
...
+#ifdef HAL_RTC_MODULE_ENABLED
+extern RTC_HandleTypeDef hrtc;
+time_t stm32_hal_time(time_t *t1)
+{
+    struct tm tm_time;
+    time_t ret;
+    RTC_TimeTypeDef time;
+    RTC_DateTypeDef date;
+
+    /* must get time and date here due to STM32 HW bug */
```
Comment (dgarske) [inline wolfcrypt/src/wc_port.c:2488]: It is the order of these call that is important / bug. Time then date.

File: wolfcrypt/test/test.c [hunk 7]
```diff
@@ -642,6 +644,8 @@ static int wolfssl_pb_print(const char* msg, ...)
         ASSERT_RESTORED_VECTOR_REGISTERS(exit(1););
     }
 #else
+    #ifndef test_pass
```
Comment (dgarske) [inline wolfcrypt/test/test.c:647]: Would an `#undef test_pass` here be better?

File: wolfcrypt/test/test.c [hunk 8]
```diff
@@ -674,9 +680,11 @@ int wolfcrypt_test(void* args)
     heap_baselineBytes = wolfCrypt_heap_peakBytes_checkpoint();
 #endif
 
+#ifndef NO_TEST_PRINTF
```
Comment (dgarske) [inline wolfcrypt/test/test.c:683]: Top of test.c has some macros for overriding printf`. Perhaps something like `#elif defined(NO_PRINTF #define printf(...) /* not available */`.

File: wolfssl/wolfcrypt/settings.h [hunk 9]
```diff
@@ -2300,6 +2300,13 @@ extern void uITRON4_free(void *p) ;
     #define SSL_CTRL_SET_TLSEXT_HOSTNAME 55
 #endif
 
+#ifdef WOLFSSL_LWIP_NATIVE
+    #include "lwip/tcp.h"
+    #include "lwip/sockets.h"
+
+    #undef WOLFSSL_USER_IO
+    #define WOLFSSL_USER_IO
```
Comment (dgarske) [inline wolfssl/wolfcrypt/settings.h:2308]: Force user IO for LWIP native?

**Review by dgarske — CHANGES_REQUESTED** [review]
Super excited to see this LWIP Native support back! Thanks Jacob. Just a couple minor things.

**Comment by JacobBarthelmeh** [comment]
Thanks for the review David, I addressed some but not all yet. Am testing on the device for the comments regarding test.c and WOLFSSL_USER_IO.

**Review by dgarske — APPROVED** [review]

**Pull request merged**